deepchem/rl/mcts.py

```diff
@@ -0,0 +1,411 @@
+      feed_dict = self._create_feed_dict(state)
+      probabilities = self._graph.session.run(self._pred_prob, feed_dict=feed_dict)
+      if deterministic:
+        return probabilities.argmax()
+      else:
+        return np.random.choice(
+            np.arange(self._env.n_actions), p=probabilities[0])
+
+  def restore(self):
+    """Reload the model parameters from the most recent checkpoint file."""
+    last_checkpoint = tf.train.latest_checkpoint(self._graph.model_dir)
+    if last_checkpoint is None:
+      raise ValueError('No checkpoint found')
+    with self._graph._get_tf("Graph").as_default():
+      variables = tf.get_collection(
+          tf.GraphKeys.GLOBAL_VARIABLES, scope='global')
+      saver = tf.train.Saver(variables)
+      saver.restore(self._graph.session, last_checkpoint)
+
+  def _create_feed_dict(self, state):
+    """Create a feed dict for use by predict() or select_action()."""
+    feed_dict = dict((f.out_tensor, np.expand_dims(s, axis=0))
+                     for f, s in zip(self._features, state))
+    return feed_dict
+
+  def _run_episodes(self, steps, temperature, saver, adapt_puct):
+    """Simulate the episodes for one iteration."""
+    buffer = []
+    self._env.reset()
+    root = TreeSearchNode(0.0)
+    for step in range(steps):
+      prob, reward = self._do_tree_search(root, temperature, adapt_puct)
+      state = self._env.state
+      if not self._state_is_list:
+        state = [state]
+      buffer.append((state, prob, reward))
+      action = np.random.choice(np.arange(self._env.n_actions), p=prob)
+      self._env.step(action)
+      if self._env.terminated:
+        self._env.reset()
+        root = TreeSearchNode(0.0)
+      else:
+#        root = root.children[action]
+        root = TreeSearchNode(0.0)
+      if time.time() > self._checkpoint_time:
+        self._checkpoint_index += 1
+        saver.save(
+            self._graph.session, self._graph.save_file, global_step=self._checkpoint_index)
+        self._checkpoint_time = time.time()
+    return buffer
+
+  def _optimize_policy(self, buffer, epochs):
+    """Optimize the policy based on the replay buffer from the current iteration."""
+    batch_size = self._graph.batch_size
+    n_batches = len(buffer) // batch_size
+    for epoch in range(epochs):
+      np.random.shuffle(buffer)
+
+      def generate_batches():
+        for batch in range(n_batches):
+          indices = list(range(batch*batch_size, (batch+1)*batch_size))
+          feed_dict = {}
+          for i, f in enumerate(self._features):
+            feed_dict[f] = np.stack(buffer[j][0][i] for j in indices)
+          feed_dict[self._search_prob] = np.stack(buffer[j][1] for j in indices)
+          feed_dict[self._search_value] = np.array([buffer[j][2] for j in indices])
+          yield feed_dict
+
+      loss = self._graph.fit_generator(generate_batches(), checkpoint_interval=0)
+
+  def _do_tree_search(self, root, temperature, adapt_puct):
```

Comment:
I'm trying to understand this algorithm a little bit better.

It looks like it basically does an alternating BFS/DFS, always starting from the root. The rewards are only recorded along the "trace", i.e. the chosen path down this tree. All other off-trace nodes have a reward of zero and do not contribute towards the root's count/rewards. Am I getting this correct?

Reply:
The term "tree search" is misleading because it *isn't* really a tree search.  It runs a set of simple rollouts and builds a tree of the states that happened to be encountered by them, but it's not doing any sort of exhaustive exploration of all branches.